File: Belly/BellyGraph/src/main/java/lk/ucsc/research/bellygraph/tcm/TCMSketch.java
package lk.ucsc.research.bellygraph.tcm;

import java.util.Arrays;
import lk.ucsc.research.bellygraph.Graph;
import lk.ucsc.research.bellygraph.Node;
import org.graphstream.graph.implementations.SingleGraph;

/**
 *
 * @author agentmilindu
 */
class TCMSketch implements Graph{

    private int[][] sketch;
    int size;
    
    private org.graphstream.graph.Graph graph2;

    public TCMSketch(int size) {
        this.size = size;
        this.sketch = new int[size][size];

        graph2 = new SingleGraph("WebGraphSketch");
        graph2.setAttribute("ui.title", "GraphSketch["+size+"]");
        graph2.setStrict(false);
        graph2.setAutoCreate(true);

        String styles = "node { fill-color: red; }";
        graph2.addAttribute("ui.stylesheet", styles);

        graph2.display();

    }

    public void addEdge(Node a, Node b, int w) {
        int aHash = a.hashCode() % this.size;
        int bHash = b.hashCode() % this.size;
        sketch[aHash][bHash] = sketch[aHash][bHash] + w;

        org.graphstream.graph.Node x2 = graph2.addNode(aHash + "");
        org.graphstream.graph.Node y2 = graph2.addNode(bHash + "");
        graph2.addEdge(aHash + "" + bHash, x2, y2).setAttribute("ui.label", sketch[aHash][bHash]);

    }

    public void addEdge(Node a, Node b) {
        addEdge(a, b, 1);
    }
    
    public void addUndirectedEdge(Node a, Node b, int w) {
        int aHash = a.hashCode() % this.size;
        int bHash = b.hashCode() % this.size;
        sketch[aHash][bHash] = sketch[aHash][bHash] + w;
        sketch[bHash][aHash] = sketch[bHash][aHash] + w;

        org.graphstream.graph.Node x2 = graph2.addNode(aHash + "");
        org.graphstream.graph.Node y2 = graph2.addNode(bHash + "");
        graph2.addEdge(aHash + "" + bHash, x2, y2).setAttribute("ui.label", sketch[aHash][bHash]);
    }
    
    public void addUndirectedEdge(Node a, Node b) {
        addUndirectedEdge(a, b, 1);
    }
    
    public int getIncommingEdgesCount(Node a) {
        int count = 0;
        for (int i = 0; i < this.size; i++) {
            count += sketch[i][a.hashCode()%this.size];
            
        }
        return count;
    }
    
    public int getOutgoingEdgesCount(Node a) {
        int count = 0;
        for (int i = 0; i < this.size; i++) {
            count += sketch[a.hashCode()%this.size][i];
            
        }
        return count;
    }

    public void print() {
        if(this.size>30) return;
        for (int[] row : this.sketch) {
            System.out.println(Arrays.toString(row));
        }
    }

    @Override
    public void addNode(Node a) {
        
    }

}

File: TCMSketch.py
class TCMSketch():
    def __init__(self, size, id=0):
        self.id = id
        self.size = size
        self.sketch = [[None] * size for _ in range(size)]

    def _addEdge(self, x, y, w=1):
        v = self.sketch[x][y]

        if v:
            self.sketch[x][y] += w
        else:
            self.sketch[x][y] = w
        return self.sketch[x][y]

    def _getEdge(self, x, y):
        return self.sketch[x][y]

    def addDEdge(self, a, b, w=1):
        x = a.hash(self.id) % self.size
        y = b.hash(self.id) % self.size

        return self._addEdge(x, y, w)

    def getEdge(self, a, b):
        x = a.hash(self.id) % self.size
        y = b.hash(self.id) % self.size

        return self._getEdge(x, y)

    def addUDEdge(self, a, b):
        x = a.hash(self.id) % self.size
        y = b.hash(self.id) % self.size
        self._addEdge(x, y)
        return self._addEdge(y, x)

    def getIncommingEdgesCount(self, a):
        x = a.hash(self.id) % self.size
        count = 0
        for y in range(self.size):
            if self.sketch[y][x]:
                count += self.sketch[y][x]
        return count

    def getOutgoingEdgesCount(self, a):
        x = a.hash(self.id) % self.size
        count = 0
        for y in range(self.size):
            if self.sketch[x][y]:
                count += self.sketch[x][y]
        return count

    def show(self):
        print('\n'.join([''.join(['{:4}'.format(item) for item in row]) for row in self.sketch]))


